### ISSUE TITLE
False positive with `__new__` inside metaclass `__call__`

### ISSUE THREAD

**Issue by patrick-kidger** [microsoft/pyright#4911]
```python
class Meta(type):
    def __call__(cls, *args, **kwargs):
        self = cls.__new__(cls, *args, **kwargs)  # Argument of type "Self@Meta" cannot be assigned to parameter "cls" of type "Type[Self@__new__]" in function "__new__"; Type "Self@Meta" cannot be assigned to type "Type[Self@__new__]" (reportGeneralTypeIssues)
        return self
```
Expected behaviour is no error.

Python 3.8.16; pyright version 1.1.302 run from the command line.

**Comment by erictraut** [comment]
Pyright's behavior here is arguably correct, at least in terms of type checking rules. Mypy generates the same error in this case.

The inferred type of `cls` in this case is `Self@Meta`, and this is not type compatible with the type of the `cls` parameter in any of the overloads of the `builtins.type.__new__` method.

**Comment by patrick-kidger** [comment]
In other words, this is an upstream bug with how `builtins.type.__new__` is annotated?

**Comment by erictraut** [comment]
I'm sympathetic to the way that the typeshed maintainers have annotated `builtins.type.__new__`.

```python
@overload
def __new__(cls, __o: object) -> type: ...

@overload
def __new__(
    cls: type[_typeshed.Self], __name: str, __bases: tuple[type, ...], __namespace: dict[str, Any], **kwds: Any
) -> _typeshed.Self: ...
```

This works in most use cases. In your use case, it doesn't work. 

I suppose that a third overload could be added to handle the case where the `cls` parameter accepts a `type` and the method returns an `object`. I'm not sure that's all that useful though.

```python
    @overload
    def __new__(
        cls: type, __name: str, __bases: tuple[type, ...], __namespace: dict[str, Any], **kwds: Any
    ) -> object: ...
```

My recommendation is to annotate the type of the `cls` parameter in the `__call__` method as `Any`.

```python
class Meta(type):
    def __call__(cls: Any, *args: Any, **kwargs: Any) -> object:
        self = cls.__new__(cls, *args, **kwargs)
        return self
```
